### ISSUE TITLE
exp: dvc.yaml scheme (checkpoint field)

### ISSUE THREAD

**Issue by jorgeorpinel** [iterative/dvc#5530]
Questions/concerns coming from the docs process on `dvc exp`-related refs (in order or imporance).

> **Still WIP but feel free to comment on the questions so far.**

### ~~BUG `dvc exp diff` isn't accepting exp names~~

EDIT: Fixed in #5543

```console
$ dvc exp show --no-pager
┏━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┓
┃ Experiment    ┃ Created      ┃
┡━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━┩
│ workspace     │ -            │
│ master        │ Feb 24, 2021 │
│ └── exp-e6db1 │ Feb 24, 2021 │
└───────────────┴──────────────┘
$ dvc exp diff exp-e6db1
ERROR: failed to show experiments diff - Invalid commit 'exp-e6db1'
```

### `checkpoint: true` field (dvc.yaml):

- This field it's at the `outs` level because checkpoint outputs get a special treatment (see https://github.com/iterative/dvc.org/pull/2200#discussion_r583384787). But at the same time this signals DVC that the whole state produces checkpoints (causes all outputs to be cached at each checkpoint). So should it be at the stage level?
- Should it be `checkpoints` (plural)? There's no use for a single checkpoint I think.
- Redesign?
  > checkpoint outs are circular dependencies... In DVC we don't have proper internal support for that, but the checkpoint out behavior is a good enough workaround for now.
  (from https://github.com/iterative/dvc.org/pull/2200#discussion_r583469473)

---
Will extract these to separate issues:

### `exp show` ordering

- By default exps are sorted chronologically, but the timestamp is not shown and it's not something you can tell from the names or hashes.
- `--sort-by` only affects experiments inside groups. Parent experiments are always sorted chronologically (could they be sorted by the sort-by criteria too?

### On `exp diff` 

- `a_rev` defaults to the workspace (OK), but when an `a_rev` is given, then `b_rev` defaults to the workspace. This is consistent with `metrics/params diff` but in the context of `exp` maybe it makes more sense to default to `HEAD`? Because presumably the given `a_rev` will most often be an experiment based on `HEAD` (visible with plan `dvc exp show`).
  And 2. maybe that would also be more intuitive for regular `metrics/params diff` anyway...
- Should it be merged with `metrics/params diff` as a single top-level command? With the current defaults they're basically aliases of the same command except they get out of sync e.g. the latter now have `--targets`. Idea: `dvc compare [--only-metrics/params]` — could be extracted to its own issue.

### `dvc.api.make_checkpoint()`

Should we provide some code samples (or even a repo/libraries) in other programming languages that perform the same function? (write the signal file, block code execution until it's deleted) E.g. in R? C? Idk

**Comment by dberenbaum** [comment]
>     * Should it be `checkpoints` (plural)? There's no use for a single checkpoint I think.

Is it intended as a verb, similar to `cache` and `persist`?
 
>     * Redesign?

It's an experimental feature for now. It may be subject to redesign in the future. A little late for 2.0!

>     * `a_rev` defaults to the workspace (OK), but when an `a_rev` is given, then `b_rev` defaults to the workspace. This is consistent with `metrics/params diff` but in the context of `exp` maybe it makes more sense to default to `HEAD`? Because presumably the given `a_rev` will most often be an experiment based on `HEAD` (visible with plan `dvc exp show`).
>       And 2. maybe that would also be more intuitive for regular `metrics/params diff` anyway...

IMHO it's most intuitive to always assume a comparison to the workspace when an argument is missing. I can see the argument otherwise for experiments, but I think that HEAD will not always be a baseline against which users are comparing (there might not be an obvious baseline set of parameters). I might also be confused about whether the experiments have actually made changes to my workspace if they don't show up when doing a `diff`.

Maybe more importantly, I don't see `exp metrics/params diff` being all that useful to spend too much time worrying about, but I could be wrong.

>     * Should it be merged with `metrics/params diff` as a single top-level command? With the current defaults they're basically aliases of the same command except they get out of sync e.g. the latter now have `--targets`. Idea: `dvc compare [--only-metrics/params]` — could be extracted to its own issue

I like the idea of merging with `metrics/params diff`, although I don't know if there's enough value to changing those existing commands. Good idea in the context of ongoing discussions about streamlining the UI.

> ### `dvc.api.make_checkpoint()`
> 
> Should we provide some code samples (or even a repo/libraries) in other programming languages that perform the same function? (write the signal file, block code execution until it's deleted) E.g. in R? C? Idk

Agree that it's probably necessary to have some kind of example to better illustrate. Not sure if it's feasible by the 2.0 release date, but maybe shortly afterwards. Unless it's likely that this will be redesigned, or if we want to wait and see if there's sufficient interest in using this feature outside of Python.

**Comment by jorgeorpinel** [comment]
Good point about "persist", but I don't think "to checkpoint" is a verb.

BTW this issue doesn't intend to get resolved by Wed.

We can extract the `diff` command merge issue I suppose. But IMO definitely if they do the exact same thing then there's no point having 3 commands. That said once merged there could be aliases setup for backward compatibility.

**Comment by jorgeorpinel** [comment]
WDYT about the level at which `checkpoint` field exists in dvc.yaml @dberenbaum?

**Comment by pmrowla** [comment]
`checkpoint` applies to specific outs within a stage, the same way that `persist` or `cache/uncached` only applies to specific outs within a stage. If we made `checkpoint` a stage level flag, we would still need some other output level flag to indicate which outs are "checkpoint-able/circular-dependency" outs.

I think it is simple enough to just say that "a checkpoint stage is a stage containing one or more outs with the `checkpoint` flag" and don't really see why the flag should be moved?

**Comment by jorgeorpinel** [comment]
> it is simple enough to just say that "a checkpoint stage is a stage containing one or more outs with the checkpoint flag"

Sure, documenting one way or another is doable (done in iterative/dvc.org/pull/2200). My Q is whether it's the best scheme and UX.

I see why `checkpoints: true` makes sense at the `outs` level like similar fields. But why is it required that a stage has one+ of these for `exp run` to "listen" to the stage code `make_checkpoint()` calls? Could these things be less coupled? I.e. `exp run` always check for checkpoint signal files AND separately treat checkpoint outs specially. And in that case maybe even rename the field to something more descriptive like `circular` (just an idea).

**Comment by jorgeorpinel** [comment]
p.s. I added a few more entries to the issue description and removed WIP from the title, please re-read guys, thanks!
(I'll open other issues if needed and consolidate later.)

**Comment by jorgeorpinel** [comment]
Oh I see you already answered some of that in https://github.com/iterative/dvc.org/pull/2200#discussion_r584406312 @pmrowla 

> > why is it required that a stage has one+ of these for exp run to "listen" to the stage code make_checkpoint() calls? 
>
> it was written this way given that this is new behavior, and it's not exactly clear what DVC should be doing for stages that don't contain explicit checkpoint outs... For checkpoint outs, the user's stage should still be reproducible in between checkpoints... For "regular" stages, the expectation is that only the full stage is reproducible.

OK it gets complicated... Need to think that through ⌛

**Comment by jorgeorpinel** [comment]
> BUG? dvc exp diff isn't accepting exp names

Hi @pmrowla were you able to check this?

```console
$ dvc exp list
master:
        exp-d662e
$ dvc exp diff exp-d662e
ERROR: failed to show experiments diff - Invalid commit 'exp-d662e'
```

> We can extract to a separate issue if you prefer.

**Comment by pmrowla** [comment]
Yeah that's a bug, it should go into a separate issue

**Comment by jorgeorpinel** [comment]
Sounds good, thanks @pmrowla !

**Comment by jorgeorpinel** [comment]
From https://github.com/iterative/dvc/discussions/5614#discussioncomment-474826:

> What about adjusting the signaling method so that the live signal file also contains the paths to the (checkpoint outputs)?
essentially this is just moving the (`checkpoint(s)` field) out of dvc.yaml and into the ... signal file
> (but)
> we need to support signaling from any arbitrary (non-python) user stages, and requiring a simple "create file, wait for file to be removed" step is much more straightforward than, "generate file containing some specifically formatted list of files that DVC should treat as checkpoint outs...

**Comment by jorgeorpinel** [comment]
I'm still wondering if we need specific checkpoint outputs @pmrowla. Why not cache all outputs on every checkpoint? (If so the `checkpoint(s)` field can go on the stage level or `exp run` can always check for signal files and avoid the field in dvc.yaml.)

> Sorry if this is getting repetitive 😬 

- "given that this is new behavior": if you don't write checkpoint signal files, nothing changes and we keep assuming that there are no circular dependency files. If you do, all outputs get cached (if they have changed) on every checkpoint and the checkpoint behavior kicks in.
- "not exactly clear what DVC should be doing for stages that don't contain explicit checkpoint outs": nothing if they don't write checkpoint signal files?

In summary I like the idea of basing this feature on checkpoint signal files. Writing those makes the state incompatible with `repro` and specific for `exp run` checkpoints. All outputs get cached on every checkpoint (run-cache is also updated I suppose?).

**Comment by jorgeorpinel** [comment]
> Why not cache all outputs on every checkpoint?

Is it b/c we don't know which ones will be used as circular dependencies in the code?

**Comment by pmrowla** [comment]
We capture the state of the entire repo on every checkpoint commit. If any outputs have changed (regardless of the type of output), we will update `dvc.lock` and capture the state of those outputs.

However, `checkpoint` outs are still treated differently

- `checkpoint` outs are not removed at the start of `exp run` (they function like`persist` outs here).
- On `exp run --reset`, `checkpoint` outs are removed entirely, regardless of whether or not there is a `dvc.lock` entry for that output (see https://github.com/iterative/dvc/issues/5553). All other output types are unaffected by `exp run --reset`.

The 2nd point is the main reason that `checkpoint` is explicitly needed (otherwise they would just be regular `persist` outs).

**Comment by pmrowla** [comment]
Also, the signal file suggestion was just a brainstorming idea. As I noted in the discussion thread, I do not actually think we should get into behavior where DVC can retroactively modify a defined stage (i.e. change `dvc.yaml`) during execution (`repro`/`exp run`). I think stage/`dvc.yaml` modification should be an explicit step, separate from running a stage or pipeline.

**Comment by dberenbaum** [comment]
@jorgeorpinel It's a little unclear what's still a concern here. We could open smaller issues for the remaining items. What do you think?

**Comment by jorgeorpinel** [comment]
Agreed, will review the conversation above and open a smaller issue if needed. Thanks

**Comment by jorgeorpinel** [comment]
@dberenbaum I guess the only 2 things were

1. the idea of changing the field name to plural: `checkpoints : true/false`
2. discussion around circular dependencies. Should there be another way to implement those besides a checkpoint output? And if so would we still need the `checkpoint` field in that case?

So nothing super actionable, more of a discussion. Do you think any of that is worth considering?

**Comment by dberenbaum** [comment]
>     1. the idea of changing the field name to plural: `checkpoints : true/false` (since the stage and output can have multiple checkpoints)

I think this makes sense, but not sure how easy it would be to make the change and make it non-breaking.


>     2\. discussion around circular dependencies. Should there be another way to implement those besides a checkpoint output? And if so would we still need the `checkpoint` field in that case?

🤔 This one might be interesting WRT ongoing discussions about how to "end" checkpoints experiments (#6104 and #6183). The user's code can specify the point at which no new checkpoints will be created, but the checkpoint stage will still always run even if it doesn't generate new checkpoints. This could create problems - what if I don't want to download the massive training data to run this stage again since I'm finished with model training? dvc could instead confirm that the output of the stage did not change from the input and skip running this stage altogether.

**Comment by jorgeorpinel** [comment]
> make the change and make it non-breaking

Seen breaking changes in minor 2.x releases. But it could be schedule for 3.0. Up to you guys

> dvc could instead confirm that the output of the stage did not change from the input [last state of the same output, I think?] and skip running this stage altogether

Good idea! Sounds like there are other issues where you may reuse the alternative if appropriate then. So I think we're done here.
